tests/server/master_server/chunk_server_manager_test.cc

```diff
@@ -17,48 +20,637 @@ class ChunkServerManagerTest : public ::testing::Test {
+  for (ushort i = 0; i < chunk_count; ++i) {
+    chunk_server->add_stored_chunk_handles(std::to_string(i));
+  }
+
+  EXPECT_TRUE(
+      ChunkServerManager::GetInstance().RegisterChunkServer(chunk_server));
+
+  EXPECT_EQ(*chunk_server,
+            ChunkServerManager::GetInstance().GetChunkServer(*server_location));
+
+  const uint32_t available_disk_mb = 100;
+
+  absl::flat_hash_set<std::string> chunks_to_add;
+  chunks_to_add.insert("10");
+  chunks_to_add.insert("11");
+
+  absl::flat_hash_set<std::string> chunks_to_remove;
+  chunks_to_remove.insert("4");
+  chunks_to_remove.insert("0");
+  chunks_to_remove.insert("3");
+  chunks_to_remove.insert("1");
+  chunks_to_remove.insert("2");
+
+  ChunkServerManager::GetInstance().UpdateChunkServer(
+      *server_location, available_disk_mb, chunks_to_add, chunks_to_remove);
+
+  // check that the chunks are updated
+  // we have removed 0 to 4 and added 10 and 11,
+  // so there should be chunks 5 to 11
+  const ushort total_chunks_count = 12;
+  for (ushort i = 0; i < total_chunks_count; ++i) {
+    auto chunk_handle = std::to_string(i);
+    auto chunk_locations =
+        ChunkServerManager::GetInstance().GetChunkLocations(chunk_handle);
+
+    if (chunks_to_remove.contains(chunk_handle)) {
+      EXPECT_TRUE(chunk_locations.empty());
+    } else {
+      EXPECT_EQ(1, chunk_locations.size());
+
+      EXPECT_TRUE(chunk_locations.contains(*server_location));
+    }
+  }
+
+  // Check that the available disk was updated
+  auto allocated_locations =
+      ChunkServerManager::GetInstance().AllocateChunkServer("TestHandle_",
+                                                            /*server_count=*/1);
+
+  EXPECT_EQ(1, allocated_locations.size());
+
+  EXPECT_TRUE(allocated_locations.contains(*server_location));
+}
+
+// Concurrency test
+
+// Test concurrent chunkserver registration, then concurrent chunkserver
+// unregister.
+TEST_F(ChunkServerManagerTest,
+       ConcurrentRegisterAndConcurrentUnregisterChunkServers) {
+  ushort thread_count = 10;
+  std::vector<std::thread> threads;
+
+  const ushort chunk_count = 100;
+
+  for (ushort i = 0; i < thread_count; i++) {
+    threads.push_back(std::thread([&, i, chunk_count]() {
+      protos::ChunkServerLocation* server_location =
+          new protos::ChunkServerLocation(std::move(CreateChunkServerLocation(
+              std::string("192.168.1.") + std::to_string(i), i)));
+
+      std::shared_ptr<protos::ChunkServer> chunk_server(
+          new protos::ChunkServer());
+      chunk_server->set_allocated_location(server_location);
+      chunk_server->set_available_disk_mb(i);
+
+      // Add same chunk handles to the chunk server
+      for (ushort i = 0; i < chunk_count; ++i) {
+        chunk_server->add_stored_chunk_handles(std::to_string(i));
+      }
+
+      ChunkServerManager::GetInstance().RegisterChunkServer(chunk_server);
+    }));
+  }
+
+  // Join all threads
+  for (ushort i = 0; i < thread_count; i++) {
```

Comment:
See tests/utils.h, there is a helper function for this/

Reply:
sure, will use that util some other time.